src/map/mob.cpp

```diff
@@ -2098,19 +2099,51 @@ static TIMER_FUNC(mob_ai_hard){
 
 /**
  * Set random option for item when dropped from monster
- * @param itm Item data
- * @param mobdrop Drop data
+ * @param item: Item data
+ * @param mobdrop: Drop data
  * @author [Cydh]
  **/
-void mob_setdropitem_option(struct item *itm, struct s_mob_drop *mobdrop) {
-	struct s_random_opt_group *g = NULL;
-	if (!itm || !mobdrop || mobdrop->randomopt_group == RDMOPTG_None)
+void mob_setdropitem_option(item *item, s_mob_drop *mobdrop) {
+	if (!item || !mobdrop)
 		return;
-	if ((g = itemdb_randomopt_group_exists(mobdrop->randomopt_group)) && g->total) {
-		int r = rnd()%g->total;
-		if (&g->entries[r]) {
-			memcpy(&itm->option, &g->entries[r], sizeof(itm->option));
-			return;
+
+	std::shared_ptr<s_random_opt_group> group = random_option_group.find(mobdrop->randomopt_group);
+
+	if (group) {
+		// Apply Must options
+		for (size_t i = 0; i < group->slot.size(); i++) {
+			std::vector<std::shared_ptr<s_random_opt_group_entry>> options = group->slot[static_cast<uint16>(i)];
+			int32 r = 100 * options.size() + rnd() % options.size(), idx = 0; // !TODO: Config the '100' to keep from getting stuck in a loop?
+
+			while (r > 0 && rnd() % 10000 >= options[idx]->chance) {
+				idx = (idx + 1) % options.size();
+				--r;
+			}
+
+			item->option[i].id = options[idx]->id;
+			item->option[i].value = rnd_value(options[idx]->min_value, options[idx]->max_value);
+			item->option[i].param = options[idx]->param;
+		}
+
+		// Apply Random options (if available)
+		if (group->max_random > 0) {
+			std::vector<std::shared_ptr<s_random_opt_group_entry>> options = group->random_option;
+
+			for (size_t i = 0; i < group->max_random; i++) {
```

Comment:
I think the inner and outer loop should be switched a little maybe or merged?

Pseudo code:
```
for (size_t i = 0; i < min( group->max_random, MAX_ITEM_RDM_OPT ); i++) {
	// If it already has a option  - skip it
	if (item->option[j].id > 0)
		continue;
	// Apply random logic
}
```

Random logic from here and above could be put into a sub function again.